experimental/iterators/lib/Conversion/IteratorsToLLVM/IteratorsToLLVM.cpp

```diff
@@ -196,6 +229,530 @@ struct PrintOpLowering : public ConversionPattern {
+                       const IteratorAnalysis::IteratorInfo &opInfo) {
+  return createOpenNextCloseFunc(originalOp, rewriter, opInfo.stateType,
+                                 opInfo.stateType, opInfo.closeFunc);
+}
+
+//===----------------------------------------------------------------------===//
+// Conversion of iterator ops to LLVM
+//===----------------------------------------------------------------------===//
+
+/// Converts the given SampleInputOp to LLVM using the converted operands from
+/// the upstream iterator.
+FailureOr<Optional<Value>>
+convertIteratorOp(SampleInputOp op, ArrayRef<Value> operands,
+                  RewriterBase &rewriter,
+                  const IteratorAnalysis &typeAnalysis) {
+  assert(op->getNumResults() == 1);
+  StreamType streamType = op->getResult(0).getType().dyn_cast<StreamType>();
+  assert(streamType);
+  Type elementType = streamType.getElementType();
+
+  auto maybeOpInfo = typeAnalysis.getIteratorInfo(op);
+  assert(maybeOpInfo.hasValue());
+  auto opInfo = maybeOpInfo.getValue();
+
+  // Create converted state.
+  Value convertedState =
+      rewriter.create<UndefOp>(op->getLoc(), opInfo.stateType);
+
+  // Create Open function. --------------------------------------------------
+  {
+    FuncOp funcOp = createOpenFunc(op, rewriter, opInfo);
+
+    Block *block = &funcOp.getBody().front();
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(block);
+
+    // Insert constant zero into state.
+    Value zero = rewriter.create<LLVM::ConstantOp>(
+        op->getLoc(), rewriter.getI32Type(), rewriter.getI32IntegerAttr(0));
+    Value updatedState =
+        rewriter.create<InsertValueOp>(op->getLoc(), block->getArgument(0),
+                                       zero, rewriter.getIndexArrayAttr({0}));
+
+    // Return.
+    rewriter.create<func::ReturnOp>(op->getLoc(), updatedState);
+  }
+
+  // Create Next function. --------------------------------------------------
+  {
+    FuncOp funcOp = createNextFunc(op, rewriter, opInfo, elementType);
+
+    Block *block = &funcOp.getBody().front();
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(block);
+
+    Value originalState = block->getArgument(0);
+    Value currentIndex = rewriter.create<ExtractValueOp>(
+        op->getLoc(), rewriter.getI32Type(), originalState,
+        rewriter.getIndexArrayAttr({0}));
+
+    // Test if we have reached the end of the range.
+    Value three =
+        rewriter.create<arith::ConstantIntOp>(op->getLoc(), /*value=*/3,
+                                              /*width=*/32);
+    Value hasNext = rewriter.create<arith::CmpIOp>(
+        op->getLoc(), arith::CmpIPredicate::sle, currentIndex, three);
+    scf::IfOp ifOp =
+        rewriter.create<scf::IfOp>(op->getLoc(), opInfo.stateType, hasNext,
+                                   /*withElseRegion=*/true);
+
+    // Fill "then" region of ifOp.
```

Comment:
Can we use a much more compact IR building form ?
E.g. mlir/lib/Conversion/VectorToSCF/VectorToSCF.cpp has 
```
    auto check = lb.create<scf::IfOp>(
        resultTypes, cond,
        /*thenBuilder=*/
        [&](OpBuilder &b, Location loc) {
          maybeYieldValue(b, loc, hasRetVal, inBoundsCase(b, loc));
        },
        /*elseBuilder=*/
        [&](OpBuilder &b, Location loc) {
          if (outOfBoundsCase) {
            maybeYieldValue(b, loc, hasRetVal, outOfBoundsCase(b, loc));
          } else {
            b.create<scf::YieldOp>(loc);
          }
        });
```

Reply:
Ah, nice pointer! I didn't know about this! Done in next commit.